efficientdet/third_party/grad_checkpoint/memory_saving_gradients.py

```diff
@@ -0,0 +1,521 @@
+  fwd_ops = ge.get_forward_walk_ops([x.op for x in xs],
+                                    inclusive=True,
+                                    within_ops=bwd_ops)
+  logging.debug("fwd_ops: %s", len(fwd_ops))
+
+  # exclude ops with no inputs
+  fwd_ops = [op for op in fwd_ops if op.inputs]
+
+  logging.debug("fwd_ops: %s", len(fwd_ops))
+  # don't recompute xs, remove variables
+  xs_ops = _to_ops(xs)
+  fwd_ops = [op for op in fwd_ops if not op in xs_ops]
+  fwd_ops = [op for op in fwd_ops if not '/assign' in op.name]
+  fwd_ops = [op for op in fwd_ops if not '/Assign' in op.name]
+  fwd_ops = [op for op in fwd_ops if not '/read' in op.name]
+  logging.debug("fwd_ops: %s", len(fwd_ops))
+  ts_all = ge.filter_ts(fwd_ops, True)  # get the tensors
+  logging.debug("ts_all: %s", len(ts_all))
+
+  ts_all = [t for t in ts_all if '/read' not in t.name]
+  ts_all = set(ts_all) - set(xs) - set(ys)
+  logging.debug("ts_all: %s", len(ts_all))
+
+  # construct list of tensors to checkpoint during forward pass, if not
+  # given as input
+  if not isinstance(checkpoints, list):
+    if checkpoints == 'collection':
+      checkpoints = tf.get_collection('checkpoints')
+
+    elif checkpoints == 'speed':
+      # checkpoint all expensive ops to maximize running speed
+      checkpoints = ge.filter_ts_from_regex(fwd_ops, 'conv2d|Conv|MatMul')
+
+    elif checkpoints == 'memory':
+
+      # remove very small tensors and some weird ops
+      def fixdims(
+          t
+      ):  # tf.Dimension values are not compatible with int, convert manually
+        try:
+          return [int(e if e is not None else 64) for e in t.as_list()]
+        except AttributeError as e:
+          logging.error("%s", e)
+          logging.error("unknown shape %s", t)
+          return [0]  # unknown shape
+
+      ts_all = [
+          t for t in ts_all
+          if np.prod(fixdims(t.shape)) > MIN_CHECKPOINT_NODE_SIZE
+          # if (tf.size(t) > MIN_CHECKPOINT_NODE_SIZE)
+      ]
+      logging.debug("ts_all: %s", len(ts_all))
+      ts_all = [t for t in ts_all if 'L2Loss' not in t.name]
+      ts_all = [t for t in ts_all if 'entropy' not in t.name]
+      ts_all = [t for t in ts_all if 'FusedBatchNorm' not in t.name]
+      ts_all = [t for t in ts_all if 'Switch' not in t.name]
+      ts_all = [t for t in ts_all if 'dropout' not in t.name]
+      # DV: FP16_FIX - need to add 'Cast' layer here to make it work for FP16
+      ts_all = [t for t in ts_all if 'Cast' not in t.name]
+      logging.debug("ts_all: %s", len(ts_all))
+
+      # filter out all tensors that are inputs of the backward graph
+      with util.capture_ops() as bwd_ops:
+        tf_gradients(ys, xs, grad_ys, **kwargs)
+
+      bwd_inputs = [t for op in bwd_ops for t in op.inputs]
+      # list of tensors in forward graph that is in input to bwd graph
+      ts_filtered = list(set(bwd_inputs).intersection(ts_all))
+      debug_print("Using tensors %s", ts_filtered)
```

Comment:
What's difference between debug_print and logging.debug.
Maybe we should use absl.logging.
Could you add some test cases for memory_saving_gradients? Thanks.

Reply:
Hi @fsx950223  
1) Regarding debug_print you can refer to the function description 
```python 
  """Like logger.log, but also replaces all TensorFlow ops/tensors with their
  names. Sensitive to value of DEBUG_LOGGING, see enable_debug/disable_debug
  Usage:
    debug_print("see tensors %s for %s", tensorlist, [1,2,3])
  """
```
If you want to log the tensors themselves debug_print is a good option (but it may be quite wordy). If no tensors standard logging.debug would do
2) Thanks for the suggestion. What does this library do? I do not want to add a dependency on any other third party library as it may lead to necessity to vendor them to this repo due to google internal rules. Then need refactoring and so on. Could be a neverending cycle 
3) Regarding testing. More testing coverage is always good. Probably I will add some tests at a later stage (maybe it should be part of a bigger integration test for this repo as gradient checkpointing is a major switch) and probably I will add some optimizations to the algorithm. However, this pull request is stable and it is on the way to be merged so unless it is a bug I don't want to make any changes

Reply:
We always use absl.logging in the repo.

Reply:
Thanks, as it is not a bug, I will rectify after merging this pull request